sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -245,7 +244,7 @@ case class BroadcastHashJoinExec(
            |boolean $found = false;
            |// generate join key for stream side
            |${keyEv.code}
-           |if ($anyNull) {
+           |if (${if (isLongHashedRelation) s"$anyNull" else s"${keyEv.value}.allNull()"}) {
```

Comment:
Does this work correctly?, e.g., the case where rewriting 4 short join keys `(short, short, short, short)` into a single long key?

Reply:
Oh you can combine multiple keys into a single long key ? I believe the reason for this special case with the long hashed relation is because the long hashed relation can only take a single long key. But it would be interesting if we can get multiple "multiple real keys packed into it".

Reply:
I think the multiple keys rewrite in HashJoinExec is symmetry, if four short is rewrite into a long key, then it's counted as single key, or other wise, it should be UnsafehashedRelation in right side, right? Since LongHashedRelation only take single key as look up input.

and the CodeGen logical of `$anyNull` should handle the multi key rewrite. But i am agreed that we should have test for those cases.

Reply:
> Oh you can combine multiple keys into a single long key ? I believe the reason for this special case with the long hashed relation is because the long hashed relation can only take a single long key. But it would be interesting if we can get multiple "multiple real keys packed into it".

FYI, it is ok to pack multi key into a Long, as long as there are IntegralType and defaultSizeSum is less or equal than 8
```
def rewriteKeyExpr(keys: Seq[Expression]): Seq[Expression] = {
    assert(keys.nonEmpty)
    // TODO: support BooleanType, DateType and TimestampType
    if (keys.exists(!_.dataType.isInstanceOf[IntegralType])
      || keys.map(_.dataType.defaultSize).sum > 8) {
      return keys
    }

    var keyExpr: Expression = if (keys.head.dataType != LongType) {
      Cast(keys.head, LongType)
    } else {
      keys.head
    }
    keys.tail.foreach { e =>
      val bits = e.dataType.defaultSize * 8
      keyExpr = BitwiseOr(ShiftLeft(keyExpr, Literal(bits)),
        BitwiseAnd(Cast(e, LongType), Literal((1L << bits) - 1)))
    }
    keyExpr :: Nil
  }
```

Reply:
Thanks for digging this, I think multi-key NAAJ would preclude the packing of multiple keys into a single long key unless of all of those sub-keys are actually non nullable ?

I think it would actually be incorrect to NOT use the key explosion of HashedRelation when there are multiple logical keys that can be independently nullable. Is there a way we can 'undo' the packing into the LongHashedRelation and switch back to a regular HashedRelation ? 

Is there a test case for this scenario of multiple nullable integral keys ?

Reply:
I still insist that we should not make it special here.

no matter it combine multi key into single long or not. Inside Probe looking logical,
I see

Single Long Key -> LongHashedRelation
Multi Key -> UnsafeHashedRelation

As long as the codegen make sure the above assumption Established，I can just go ahead with my NAAJ logical

Reply:
Okay lets add tests here for @viirya's case of multiple short keys packing into a single long. Please ensure that at least two of those keys are nullable in your test. I am curious if that will trigger the LongHashedRelation or the (exploding) UnsafeHashedRelation. I believe that the former is incorrect.

Basically I am claiming that LongHashedRelation cannot be used with more than one null keys. This was moot earlier because we were dropping rows that had any null key.

Reply:
At least, as @agrawaldevesh suggested in https://github.com/apache/spark/pull/29304#discussion_r463936613, I think its better to leave some comments about how it work here and we need to carefully write a logic for this code part.

Reply:
> At least, as @agrawaldevesh suggested in [#29304 (comment)](https://github.com/apache/spark/pull/29304#discussion_r463936613), I think its better to leave some comments about how it work here and we need to carefully write a logic for this code part.

agree.